include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,200 @@
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef DHT_STORAGE_HPP
+#define DHT_STORAGE_HPP
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include <string>
+
+#include <libtorrent/config.hpp>
+#include <libtorrent/time.hpp>
+#include <libtorrent/socket.hpp>
+#include <libtorrent/sha1_hash.hpp>
+#include <libtorrent/address.hpp>
+#include <libtorrent/session_settings.hpp>
+#include <libtorrent/performance_counters.hpp>
+#include <libtorrent/kademlia/item.hpp>
+
+namespace libtorrent {
+namespace dht
+{
+	// The DHT storage interface is a pure virtual class that can
+	// be implemented to customize how the data for the DHT is stored.
+	//
+	// The default storage implementation uses three maps in RAM to save
+	// the peers, mutable and immutable items and it's designed to
+	// provide a fast and fully complaint behavior of the BEPs.
+	//
+	// libtorrent comes with one built-in storage implementation:
+	// ``dht_default_storage`` (private non-accessible class). Their
+	// constructor function is called dht_default_storage_constructor().
+	//
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+#ifndef TORRENT_NO_DEPRECATE
+		// This function returns the number of torrents tracked by
+		// the DHT at the moment. It's used to fill session_status.
+		// It's deprecated.
+		//
+		virtual size_t num_torrents() const = 0;
+
+		// This function returns the sum of all of peers per torrent
+		// tracker byt the DHT at the moment.
+		// It's deprecated.
+		//
+		virtual size_t num_peers() const = 0;
+#endif
+
+		// This functions retrieve the peers tracked by the DHT
+		// corresponding to the given info_hash. You can specify if
+		// you want only seeds and/or you are scraping the data.
+		//
+		// For future implementers:
+		// If the torrent tracked contains a name, such a name
+		// must be stored as a string in peers["n"]
+		//
+		// If the scrape parameter is true, you should fill these keys:
+		// peers["BFpe"] - with the standard bit representation of a
+		//                 256 bloom filter containing the downloaders
+		// peers["BFsd"] - with the standard bit representation of a
+		//                 256 bloom filter containing the seeders
+		//
+		// If no scraping, you should fill the key peers["values"] with
+		// a list containing a subset of peers tracked by the given
+		// info_hash. Such a list should consider the value of
+		// dht_settings::max_peers_reply. If noseed is true only peers
+		// marked as no seed should be included.
+		//
+		// returns true if an entry with the info_hash is found and
+		// the data is returned inside the (entry) out parameter peers.
+		//
+		virtual bool lookup_peers(sha1_hash const& info_hash
```

Comment:
I wonder if this function should be called get_peers() (for symmetry with the protocol)